### ISSUE TITLE
Concurrent requests to `SUNSUBCRIBE` causes a race condition bug (`ClientClosedError`)

### ISSUE THREAD

**Issue by BrentLayne** [redis/node-redis#2685]
## Bug description
Making concurrent requests to `SUNSUBSCRIBE` causes one or more of the requests to throw the [ClientClosedError](https://github.com/redis/node-redis/blob/f4680f0849a0516453bcab1a23785cbb2888f287/packages/client/lib/errors.ts#L21-L25)

## Why is this happening?
This bug is caused by a race condition due to how the [executeShardedUnsubscribeCommand](https://github.com/redis/node-redis/blob/3bd28a9924d9675b0df76063d2fc59ca1bd70120/packages/client/lib/cluster/cluster-slots.ts#L606-L620) function works:
- The function awaits multiple async calls
- THEN the function calls `client.disconnect()`

Every time the `executeShardedUnsubscribeCommand` function encounters an `await`, function execution is paused and the javascript thread begins to process the other concurrent requests to `SUNBSUBSCRIBE`. As a result, the requests to `SUNSUBSCRIBE` are now racing against each other.

The request that 'wins' this race will call `disconnect()` first and disconnect the socket. The subsequent requests will try to disconnect the socket _again_, resulting in the `ClientClosedError` [being thrown](https://github.com/redis/node-redis/blob/68d835d7a228cc1cf6b0a18b960fbdbb2cce2e4a/packages/client/lib/client/socket.ts#L255-L257)

## What is the expected behaviour?
I can make concurrent requests to `SUNSUBSCRIBE` and all the requests resolve with no errors thrown

## Steps to reproduce

Run the below script to reproduce the race condition bug:

```javascript
import { createCluster } from 'redis'


const client = createCluster({ rootNodes });

await client.connect();

const onMessage = () => {};

await client.sSubscribe('test', onMessage);
await client.sSubscribe('test2', onMessage);

try {
  // concurrent requests to sUnsubscribe are racing against each other to disconnect the socket
  await Promise.all([
    client.sUnsubscribe('test', onMessage),
    client.sUnsubscribe('test2', onMessage),
  ]);
} catch (error) {
  // this log statement logs the ClientClosedError
  console.log('*** error ***', error);
}
```

### Node.js Version

18.18.2

### Redis Server Version

7.0.7

### Node Redis Version

4.6.7

### Platform

macOS

**Comment by BrentLayne** [comment]
I think the fix is as simple as checking if `client.isOpen` before calling `client.disconnect()` in `executeShardedUnsubscribeCommand`

i.e. the function definition should be changed to be:

```
async executeShardedUnsubscribeCommand(
        channel: string,
        unsubscribe: (client: RedisClientType<M, F, S>) => Promise<void>
    ): Promise<void> {
        const { master } = this.slots[calculateSlot(channel)];
        if (!master.pubSubClient) return Promise.resolve();

        const client = await master.pubSubClient;
        await unsubscribe(client);
        
        // NEW: Confirm that the client.isOpen before disconnecting
        if (!client.isPubSubActive && client.isOpen) {
            await client.disconnect();
            master.pubSubClient = undefined;
        }
    }
```

With this change, I no longer see an error

**Comment by BrentLayne** [comment]
Also worth noting that cluster `UNSUBSCRIBE` likely has the exact same problem because the code for [executeUnsubscribeCommand](https://github.com/redis/node-redis/blob/3bd28a9924d9675b0df76063d2fc59ca1bd70120/packages/client/lib/cluster/cluster-slots.ts#L559-L569) is structured the excat same way

**Comment by leibale** [comment]
@BrentLayne Good catch! Want to open a PR (preferably with a test)? Or you prefer I'll take it from here?

**Comment by BrentLayne** [comment]
@leibale I wrote a PR with a test to fix the issue! I'm not able to push my branch to github though because I don't have permissions:

```
ERROR: Permission to redis/node-redis.git denied to BrentLayne.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
```

Sorry if this is a dumb question, I haven't done much contributions to OSS --> do you need to give me permissions for me to make the PR?

**Comment by BrentLayne** [comment]
@leibale for your consideration: https://github.com/redis/node-redis/pull/2687

**Comment by leibale** [comment]
fixed in #2687
